code/modules/projectiles/guns/syringe_gun.dm

```diff
@@ -95,3 +94,285 @@
+		// Otherwise, we wouldn't be able to fully empty the gun's chems without firing it.
+		if(chambered?.BB && chambered.BB.reagents.total_volume)
+			transfer_amount = chambered.BB.reagents.trans_to(destination, get_units_per_shot())
+			transfer_source = "the syringe chambered in [src]"
+		else
+			transfer_amount = reagents.trans_to(destination, reservoir_transfer_amount)
+
+		to_chat(user, "<span class='notice'>You drain [transfer_amount] unit\s from [transfer_source] into [destination].</span>")
+		// Refill the syringe
+		update_loaded_syringe()
+		return TRUE
+	else
+		return ..()
+
+// Switch the amount of reagents used per shot.
+/obj/item/gun/syringe/rapidsyringe/AltClick(mob/living/user)
+
+	transfer_amount_selection++
+	if(transfer_amount_selection > length(possible_transfer_amounts))
+		transfer_amount_selection = 1
+
+	playsound(src, 'sound/weapons/gun_interactions/selector.ogg', 25, 1)
+	to_chat(user, "<span class='notice'>[src] will now fill each syringe with up to [get_units_per_shot()] units.</span>")
+	update_loaded_syringe()
+
+/// Play an alert when the reservoir has run out of reagents so people don't unknowingly dump empty syringes into their foes
+/// Running out of syringes is just handled by *click*
+/obj/item/gun/syringe/rapidsyringe/afterattack(atom/target, mob/living/user, flag, params)
+	. = ..()
+	if(!alarmed && !reagents.total_volume && (!chambered?.BB || !chambered.BB.reagents.total_volume))
+		playsound(loc, 'sound/weapons/smg_empty_alarm.ogg', 25, 1, frequency=60000)
+		to_chat(user, "<span class='userdanger'>[src] beeps: Internal reservoir empty!</span>")
+		alarmed = TRUE
+
+/// Update the chambered syringe's contents based on the reservoir contents.
+/// Makes sure that what's contained in the syringe is representative of the mix as a whole.
+/obj/item/gun/syringe/rapidsyringe/proc/update_loaded_syringe()
+	if(!chambered?.BB)
+		return
+
+	// Refresh the mix in the syringe by pushing it into our internal beaker, and then pulling it out.
+	// The only time we /won't/ is when we dump out to a beaker.
+	if(chambered.BB.reagents.total_volume)
+		chambered.BB.reagents.trans_to(src, chambered.BB.reagents.total_volume)
+
+	reagents.trans_to(chambered.BB, get_units_per_shot())
+
+/obj/item/gun/syringe/rapidsyringe/process_chamber()
+
+	if(!length(syringes) || chambered?.BB)
+		return
+
+	var/obj/item/reagent_containers/syringe/S = syringes[1]
+	if(!S)
+		return
+
+	chambered.BB = new S.projectile_type(src)
+	update_loaded_syringe()
+	chambered.BB.name = S.name
+
+	syringes.Remove(S)
+	qdel(S)
+
+// Unload an empty syringe, making sure its existing contents get returned to the reservoir
+/obj/item/gun/syringe/rapidsyringe/attack_self(mob/living/user)
+	if(!length(syringes) && !chambered.BB)
+		to_chat(user, "<span class='notice'>[src] is empty.</span>")
+		return FALSE
+
+	var/obj/item/reagent_containers/syringe/S
+	if(chambered.BB) // Remove the chambered syringe first
```

Comment:
Shouldn't matter in the grand scheme of things, but i would expect it'd be easier (IC) to remove non-chambered syringes first.
You gotta write the code anyway, but at least that would make us do fewer reagent transfers...
And maybe a slight extra reason to have a magazine.

Reply:
Yeah, I think this was again more code that just kind of came over from the regular syringe gun, and acts like other guns (cocking the gun, for instance, to pop out the chambered bullet).

Reply:
Applied this, you now unload extra syringes first.